### PR TITLE
[MJAVADOC-662] The fix operation wrongly generate illegal javadoc for static final fields.

### PR DESCRIPTION
**Pull request by XenoAmess**
Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MJAVADOC) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MJAVADOC-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MJAVADOC-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify -Prun-its` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [x] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

### CODE REVIEW

**Comment by XenoAmess** [comment]
@elharo 
fix done.
Though this one is not that severe bug, as it will not modify from correct to fault, but will only from missiing to fault.

**Comment by XenoAmess** [comment]
notice that the class StringEscapeUtils is deprecated in commons-lang3, because they want to move that class to commons-text when lang4.
But it is usable for now (if not upgrade to lang4, which is not released yet).

And I don't think it worthy for adding a additional dependency commons-text.

If you have different ideas about this, feel free to ask me change to add commons-text as dependency&use that class in it.

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractFixJavadocMojo.java [hunk 1]
```diff
@@ -1412,11 +1413,11 @@ private void addDefaultFieldComment( final StringWriter stringWriter, final Java
                 // CHECKSTYLE_OFF: MagicNumber
                 if ( value.length() < 40 )
```
Comment (elharo) [inline src/main/java/org/apache/maven/plugins/javadoc/AbstractFixJavadocMojo.java:1414]: should this check the escaped length instead now, which could be longer?

**Comment by XenoAmess** [comment]
@elharo 
> Please do use commons.text instead here.

Hi.
Then we will use commons-text 1.3, as 1.4 or later requires java 8 as runtime, and 1.3 is the last version who release on java 1.7.

details: http://commons.apache.org/proper/commons-text/changes-report.html

File: src/main/java/org/apache/maven/plugins/javadoc/AbstractFixJavadocMojo.java (hunk 1 above)
Comment (XenoAmess) [reply src/main/java/org/apache/maven/plugins/javadoc/AbstractFixJavadocMojo.java:1414]: escaped length should be >= original length (if my html knowledges are correct).
I'm OK to change it to escaped length.

File: src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java [hunk 2]
```diff
@@ -451,6 +451,21 @@ public void testException3()
     {
         throw new Exception();
     }
+
+    /**
+     * to test if it will handle static final field correctly.
+     */
+    public static final int TEST_STATIC_FINAL_FIELD_0 = 1 << 2;
```
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: @elharo 
Actually, not very surprising, especially when in a class who dealing with some bit operations.

for example:

https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/util/ConcurrentReferenceHashMap.java#L73

https://github.com/spring-projects/spring-framework/blob/master/spring-core/src/main/java/org/springframework/asm/Frame.java#L145

**Comment by XenoAmess** [comment]
@elharo Hi. Any news for this pr? Should we merge it or there still be some things that I should do for now?

File: src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (elharo) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: Yes, they're legal, but what does this add over a simple int? If I read this right, this is simply `4` (and if I'm not reading it right that just proves it's confusing)

[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: 1<<4 is simple but think about 1<<16.

[the remark this answers is not in the document]
Comment (elharo) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: Why? That's not what's in this code, and even if it were, my comment would be the same. I'm confused and surprised by this code, so I can't approve it.

**Review by elharo — CHANGES_REQUESTED** [review]

File: src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: @elharo notice that this is the test code to show in what situation the original javadoc-plugin can fail.

**Comment by XenoAmess** [comment]
@elharo notice that this is the test file to show in what situation the original javadoc-plugin (before this commit) can fail.

File: src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (elharo) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: If there's some reason it has to be this way, you need a comment explaining that. It is not self-evident. The current comment says it's testing a static final field. Static final fields don't have to have shift operators.

[the remark this answers is not in the document]
Comment (XenoAmess) [reply src/test/resources/unit/fix-test/expected/src/main/java/fix/test/ClassWithJavadoc.java:458]: > If there's some reason it has to be this way, you need a comment explaining that
got it. will do it now.

**Comment by XenoAmess** [comment]
@elharo So is there any other thing I should do now for this pr?
If nope, then I would go to bed.

**Comment by XenoAmess** [comment]
@elharo please review,thanks

**Review by elharo — COMMENTED** [review]
Running this through jenkins: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-javadoc-plugin/job/MJAVADOC-662/

**Comment by XenoAmess** [comment]
> Running this through jenkins: https://ci-builds.apache.org/job/Maven/job/maven-box/job/maven-javadoc-plugin/job/MJAVADOC-662/

thanks!

**Comment by elharo** [comment]
build failed, but it's unstable. Trying again...

**Comment by XenoAmess** [comment]
> build failed, but it's unstable. Trying again...

@elharo I saw it fails on two environments on jenkins… really weird. Honestly I'm confused.
Have you any clue for tracking this & correcting this?
Thanks.

**Comment by elharo** [comment]
Jenkins is an absolute mess. I'd like to chuck it, but I don't have that power. Those nodes are probably broken/flaky. Let me try again.

**Comment by rfscholte** [comment]
Merged with https://github.com/apache/maven-javadoc-plugin/commit/e7a3fe0ed8d7be61de9cc458751b5a24f640fba8
Thanks for the PR!

**Comment by jira-importer** [comment]
Resolve #1041
